brainiak/isfc.py

```diff
@@ -1,226 +1,1104 @@
+            if summary_statistic == np.mean:
+                isc_sample = np.tanh(np.mean(np.arctanh(isc_flipped), axis=0))
+            elif summary_statistic == np.median:
+                isc_sample = np.median(isc_flipped, axis=0)            
+        
+        # If two groups, set up group matrix get the observed difference
+        elif n_groups == 2:
+
+            # Shuffle the group assignments
+            if exact_permutations:
+                group_shuffler = np.array(exact_permutations[i])
+            elif not exact_permutations and pairwise:
+                group_shuffler = prng.permutation(np.arange(
+                    len(np.array(group_assignment)[sorter])))
+            elif not exact_permutations and not pairwise:
+                group_shuffler = prng.permutation(np.arange(
+                    len(group_assignment)))
+            
+            # If pairwise approach, convert group assignments to matrix
+            if pairwise:
+                
+                # Apply shuffler to group matrix rows/columns
+                group_shuffled = group_matrix[group_shuffler, :][:, group_shuffler]
+                
+                # Unsort shuffled matrix and squareform to create selector
+                group_selector = squareform(group_shuffled[unsorter, :][:, unsorter],
+                                            checks=False)
+                
+            # Shuffle group assignments in leave-one-out two sample test
+            elif not pairwise:
+                
+                # Apply shuffler to group matrix rows/columns
+                group_selector = np.array(group_assignment)[group_shuffler]
+                
+            # Get difference of within-group summary statistics
+            # with group permutation        
+            if summary_statistic == np.mean:
+                isc_sample = (np.tanh(np.mean(np.arctanh(
+                    iscs[group_selector == group_labels[0], :]), axis=0)) - 
+                            np.tanh(np.mean(np.arctanh(
+                    iscs[group_selector == group_labels[1], :]), axis=0)))
+            elif summary_statistic == np.median:
+                isc_sample = (np.median(
+                    iscs[group_selector == group_labels[0], :], axis=0) - 
+                            np.median(
+                    iscs[group_selector == group_labels[1], :], axis=0))
+        
+        # Tack our permuted ISCs onto the permutation distribution
+        distribution.append(isc_sample) 
+        
+        # Update random state for next iteration
+        if not exact_permutations:
+            random_state = np.random.RandomState(prng.randint(0, 2**32 - 1))
+        
+    # Convert distribution to numpy array
+    distribution = np.array(distribution)
+    assert distribution.shape == (n_permutations, n_voxels)
+
+    # Get p-value for actual median from shifted distribution
+    p = ((np.sum(np.abs(distribution) >= np.abs(observed), axis=0) + 1) /
```

Comment:
1. Can we assign a variable name for `float((len(distribution) + 1)))[np.newaxis, :]` to make this line more readable? 
2. This piece of code for p-value computation is used 3 times. Should we pack this as a function?

Reply:
1. I think `float()` was my python2 brain thinking about division, but this denominator should equal `n_permutations + 1`.
2. We should—I'll create a separate function for this.
3. Probably doesn't need to be; was just being overzealous about enforcing a particular geometry.